cupy/linalg/_eigenvalue.py

```diff
@@ -88,7 +97,56 @@ def _syevd(a, UPLO, with_eigen_vector, overwrite_a=False):
     return w.astype(w_dtype, copy=False), v.astype(v_dtype, copy=False)
 
 
-# TODO(okuta): Implement eig
+def _geev(a, with_eigen_vector, overwrite_a=False):
+    from cupy_backends.cuda.libs import cusolver
+
+    if runtime.is_hip:
+        raise NotImplementedError("geev is not implemented for HIP")
+
+    dtype, _ = _util.linalg_common_type(a)
+    _check_dtype(dtype)
+    complex_dtype = numpy.dtype(dtype.char.upper())
+
+    # Force complex-number computation for human-readable output format
+    a_ = a.astype(complex_dtype, order='F', copy=not overwrite_a)
+
+    m, lda = a.shape
+    w = cupy.empty(m, complex_dtype)
+    # Used for both right and (uncomputed) left eigenvectors
+    v = cupy.empty_like(a, dtype=complex_dtype, order='F')
+    dev_info = cupy.empty((), numpy.int32)
+    handle = device.Device().cusolver_handle
+
+    if with_eigen_vector:
+        jobvr = cusolver.CUSOLVER_EIG_MODE_VECTOR
+    else:
+        jobvr = cusolver.CUSOLVER_EIG_MODE_NOVECTOR
+    # Skip computing left eigenvectors
+    jobvl = cusolver.CUSOLVER_EIG_MODE_NOVECTOR
+
+    type_complex = _dtype.to_cuda_dtype(complex_dtype)
+    params = cusolver.createParams()
+    try:
+        work_device_size, work_host_size = cusolver.xgeev_bufferSize(
+            handle, params, jobvl, jobvr, m, type_complex, a_.data.ptr, lda,
+            type_complex, w.data.ptr, type_complex, v.data.ptr, lda,
+            type_complex, v.data.ptr, lda, type_complex)
+        work_device = cupy.empty(work_device_size, 'b')
+        work_host = numpy.empty(work_host_size, 'b')
+        cusolver.xgeev(
+            handle, params, jobvl, jobvr, m, type_complex, a_.data.ptr, lda,
+            type_complex, w.data.ptr, type_complex, v.data.ptr, lda,
+            type_complex, v.data.ptr, lda, type_complex, work_device.data.ptr,
+            work_device_size, work_host.ctypes.data, work_host_size,
+            dev_info.data.ptr)
+    finally:
+        cusolver.destroyParams(params)
+    cupy.linalg._util._check_cusolver_dev_info_if_synchronization_allowed(
+        cusolver.xgeev, dev_info)
+
+    if all(w.imag == 0.0):
```

Comment:
I don't like the [NumPy behavior](https://github.com/numpy/numpy/blob/e7a123b2d3eca9897843791dd698c1803d9a39c2/numpy/linalg/_linalg.py#L1474) but it looks like we have to follow... 🙁 

Q: can we find a way to not create an intermediate array of bools for later reduction? Perhaps this can be replaced by a simple `ReductionKernel`?